presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.execution;
+
+import com.facebook.presto.execution.buffer.OutputBufferMemoryManager;
+import com.facebook.presto.spark.classloader_interface.PrestoSparkRow;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoSparkRowBuffer
+{
+    private final OutputBufferMemoryManager memoryManager;
+
+    private final Object monitor = new Object();
+    @GuardedBy("monitor")
+    private final Queue<PrestoSparkRow> buffer = new ArrayDeque<>();
+    @GuardedBy("monitor")
+    private volatile boolean finished;
```

Comment:
Yes. Since the variable is guarded by lock the volatile modifier is not needed, as the synchronized block guarantees the ordering. However it is generally a good practice to keep the `volatile` modifier as it doesn't introduce any performance regressions while allowing to use the variable directly without synchronization where compare-and-set semantics are not needed.

Reply:
I think for boolean/byte/integer, JMM grantees it's an atomic update: https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.7 . 

@arhimondr : Are you more worried about the "Happens Before" memory model ? (See references in https://github.com/prestodb/presto/pull/8772#issuecomment-323800658)  . In that case adding `volatile` guarantees sequentially consistent  memory model :)

Reply:
Since this variable is never accessed outside the lock it doesn't have to be volatile as synchronization enforces the memory ordering. It can be safely removed. I added it just in case somebody accidental uses it outside the lock (e.g.: simple atomic read). However now I'm not sure how much of sense does it make. Let me simply remove it to avoid confusion.